pkg/ipamd/ipamd.go

```diff
@@ -1186,6 +1374,79 @@ func (c *IPAMContext) verifyAndAddIPsToDatastore(eni string, attachedENIIPs []*e
 	return seenIPs
 }
 
+// verifyAndAddPrefixesToDatastore updates the datastore with the known Prefixes. Prefixes who are out of cooldown gets added
+// back to the datastore after being verified against EC2.
+func (c *IPAMContext) verifyAndAddPrefixesToDatastore(eni string, attachedENIIPs []*ec2.Ipv4PrefixSpecification, needEC2Reconcile bool) map[string]bool {
+	var ec2VerifiedAddresses []*ec2.Ipv4PrefixSpecification
+	seenIPs := make(map[string]bool)
+	for _, privateIPv4 := range attachedENIIPs {
+		strPrivateIPv4 := aws.StringValue(privateIPv4.Ipv4Prefix)
+		log.Debugf("Check in coolddown Found prefix %s", strPrivateIPv4)
+
+		// Check if this Prefix was recently freed
+		_, ipv4CidrPtr, err := net.ParseCIDR(strPrivateIPv4)
+		if err != nil {
+			log.Debugf("Failed to parse so continuing with next prefix")
+			continue
+		}
+		found, recentlyFreed := c.reconcileCooldownCache.RecentlyFreed(strPrivateIPv4)
+		if found {
+			log.Debugf("found in cooldown")
+			if recentlyFreed {
+				log.Debugf("Reconcile skipping IP %s on ENI %s because it was recently unassigned from the ENI.", strPrivateIPv4, eni)
+				continue
+			} else {
+				if needEC2Reconcile {
+					// IMDS data might be stale
+					log.Debugf("This IP was recently freed, but is now out of cooldown. We need to verify with EC2 control plane.")
+					// Only call EC2 once for this ENI
+					if ec2VerifiedAddresses == nil {
```

Comment:
instead of check `ec2VerifiedAddresses == nil`, we should have a bool variable tracks whether GetIPv4PrefixesFromEC2 have been called or not.
otherwise, this logic's correctness relies on whether GetIPv4PrefixesFromEC2 returns empty slice or nil slice when no prefix found. we shouldn't have such dependencies in golang. (empty slice/nil slice should be treated semantically equal)

Reply:
Discussed with Yang. For now will keep it similar to secondary IPs and the right way will be to split the logic into two sections which will first fetch from EC2 once if needed and then compare with what we have from IMDS. Will update the comment.